proto/tendermint/abci/types.proto

```diff
@@ -117,27 +122,60 @@ message RequestApplySnapshotChunk {
   string sender = 3;
 }
 
+message RequestPrepareProposal {
+  bytes                   hash                 = 1;
+  tendermint.types.Header header               = 2 [(gogoproto.nullable) = false];
+  // txs is an array of transactions that will be included in a block,
+  // sent to the app for possible modifications.
+  repeated bytes          txs                  = 3;
```

Comment:
This is a meta-question, not a change request for this PR:

I'm concerned about the size of the data in a repeated transaction field. Do we have any language in the spec around how the implementation is permitted to handle corner cases, report errors, etc.? 

I expect the intent is that not all the txn for a block would be prepared at the same time, similarly to how we CheckTx one-by-one today. Given the tendency of app developers to rely on any corner case not otherwise defined, however, I wonder if we can't make some provision in the spec (e.g., "the implementation is permitted to limit the number/size of txn in this field to <some reasonable rule> provided <minimum standard>" or words to that effect.

Relatedly (and also not necessarily for this PR): We might want to pre-emptively pull this out into a message, e.g.,

```proto
message TransactionBatch {
   repeated bytes txs = 1;
}
```

That makes it easier for us to inject a layer of indirection if we later need to handle really big sets, without breaking the wire format of the enclosing message.

Reply:
The short answer is that *max_tx_bytes* below is the limit (in bytes) of the whole `txs` field. But I understand your concern as, currently, ABCI never sends/receives transactions in a batch... so this will be something we'll need to test extensively.

As you may have already figured out, I have a limited experience with protocol buffers, so if you think that wrapping a potentially huge parameter in its own message offers an extra layer of safety/security/robustness/peace of mind... then we can do definitely do that...

... and if we do it, the moment is now :-)

Reply:
Adding a message doesn't really mitigate the problem at all—it's more a question of future-proofing. 

If, for example, we wanted to include additional information in the request like "there are more transactions coming" or "I truncated this message" or whatever, that's easier to do if there's a message. 

That said: I'm not advocating this as a best practice, just thinking about how we'd adapt if we run into a situation where there is "too much" tx data. Do we drop the request on the floor? Do we truncate it silently? If so, which data do we drop? Does the application have any way to discover we did this?  Those kinds of questions.

I don't have a strong opinion here.

Reply:
I see your point. I think it's hard to say at this point. Let's leave it as it is for now, but let's also keep this in mind as we progress with the implementation.

Reply:
My idea was that if the application sent too much evidence we would log an error and the proposer wouldn't propose anything for that round. Given that Tendermint is the client and the application the server, it's hard for Tendermint to actually report errors through ABCI. I can't recall it actually happening anywhere. 

The other option which would be a lot more aggressive is that Tendermint panics when it gets back a tx set which exceeds the max bytes provided. We do a similar thing in the case of validator updates. If the application returns a validator set voting power that is greater than maxInt64 / 8, Tendermint will panic.

Reply:
Quite frankly, I would go for the second option. We are providing the App with `max_tx_bytes` in this message, so if the App returns a transaction set that is bigger than that, it's an obvious bug in the App. I don't think it's wise to try to carry on with such a bug in the App. But let's decide what to do when implementing the logic inside Tendermint.